Playwright: Part 1. Implement components, pages and flows and migrate part of wp-likes spec.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This change implements multiple components, pages and flows in Playwright.

- pages and components, where appropriate, can inherit from a new `BaseContainer` (analogous to the `AsyncBaseContainer` in Selenium).
- a basic page object for GutenbergEditor.
- components and flows to support paths touched on by the `wp-likes-spec`.

#### Testing instructions

Ensure tests pass on TeamCity for Playwright (trigger manually).

Pull the changes locally and ensure they run with `mocha --config .mocharc_playwright.yml specs/specs-playwright/wp-likes-spec.js`.

Related to #53168
Child: #53291

worldomonation committed:
Add the Playwright version of `AsyncBaseContainer`, renamed to `BaseContainer`.
- implements some of the basic functions the Selenium version of `AsyncBaseContainer` offers.

worldomonation committed:
Implement components for Likes and Navbar.

worldomonation committed:
Implement GutenbergEditorPages and NewPostFlow.

worldomonation committed:
Housekeeping changes.
- Additional JSDoc on all files.
- add partially migrated test file `wp-likes-spec`.
- add new function to generate text for the post and a TypeScript declaration for the asana package which is not in TS.

worldomonation commented:
I am guessing that as per p1621762347012600/1621505612.010500-slack-C01QG4Y91RR I will need to ensure each of the last `await` calls do a `return await` instead?

To be honest, I am still a bit confused at the rule/expectation/whatever surrounding the `return await`. Whenever I run into issues with a failing test in Playwright, I do not have issues with the calls not being in the stack.

worldomonation commented:
It also looks like the test is failing at the first step, so time to dig into why.

worldomonation committed:
Update BaseContainer to accept child class functions that do not exist in parent class.

worldomonation commented:
Looks like my tests are failing mostly because:
- the `write` link in the `masterbar/navbar` is not being found, failed to be clicked, or both of the above.
- my logic for detecting if the `like/unlike` was successful was causing issues.

worldomonation committed:
Additional promises in clickNewPost.

WunderBart commented on packages/calypso-e2e/src/lib/components/navbar-component.ts:
```diff
@@ -0,0 +1,51 @@
+	// Selectors
+	navBarSelector = '.masterbar';
+	newPostButtonSelector = '.masterbar__item-new';
+	newPostContentSelector = '.masterbar__item-new';
```
These two above have the same value. Is that intended?

worldomonation replied:
Oops, that's my mistake!

WunderBart commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -0,0 +1,106 @@
+	 */
+	async _visitPublishedPost(): Promise< void > {
+		await Promise.all( [
+			this.frame.click( this.snackBarNoticeLinkSelector ),
```
Just wanted to point out here, that the snackbar link was an unreliable (very flaky) way of getting to the published post with Selenium. It was caused by the fact that the snackbar is an animated element and is visible only for a few seconds. In [one of the recent refactors ](https://github.com/Automattic/wp-calypso/pull/52089/commits/9388bd18dae42342ee0b72e07584226c75be1c30#diff-b7f8735b885b065dc18c161fbf8de3c5bfecafd5fd04319842a8eea5f5c32180R74) I replaced the snackbar click with the `View Post` button from the post-publish menu, which fixed the flakiness: 


<p align="center">
<img src='https://user-images.githubusercontent.com/1451471/119643150-b54c4100-be1b-11eb-9831-83e60083cbcb.png' width="240" />
</p>

Another thing here is that `_visitPublishedPost` might not serve well as an independent method. It's highly dependant on the `publish` method and can be called only at a very specific moment.

worldomonation replied:
Aye, I was attempting to stay true to the original Selenium way of doing things to minimize confusion, but it seems that consideration isn't really necessary. 

It should be trivial to rewrite the visit post function to use the `View Post` button that appears in the sidebar. I will likely rename the function to something more constrained eg. `visitEntryFromSidebar` - note the removal of the word 'post' since the DOM uses the term 'entry' as a generic, and the addition of the word 'sidebar' specifically stating this method goes through the sidebar.

worldomonation replied:
Replaced the use of `snackbar` with what I call `postPublishPane`.

WunderBart commented on test/e2e/specs/specs-playwright/wp-likes-spec.js:
```diff
@@ -0,0 +1,56 @@
+import config from 'config';
+import {
+	BrowserHelper,
+	DataHelper,
```
Why do we export/import those 2 above in the PascalCase? I think we should reserve that for class names and constructors to avoid confusion.

worldomonation replied:
I do not recall in which PR, but in one of the very early PRs the convention was set.

[This StackOverflow](https://stackoverflow.com/questions/29601960/typescript-use-pascalcasing-or-camelcasing-for-module-names) post is a bit old but it appears the convention is to use PascalCase for modules and classes?

WunderBart replied:
Thanks for the context, makes sense to me! I guess the most important thing is to stay consistent 👍

WunderBart commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -0,0 +1,106 @@
+	 * @param {boolean} visit Whether to then visit the page.
+	 * @returns {Promise<void} No return value.
+	 */
+	async publish( visit = false ): Promise< void > {
```
Can we pass the `visit` attr as an object prop? It's easier to follow when calling `publish( { visit: true } )` vs `publish( true )`.

worldomonation replied:
Should not be a problem!

WunderBart commented on test/e2e/specs/specs-playwright/wp-likes-spec.js:
```diff
@@ -0,0 +1,56 @@
+		let gEditor;
+		let likesComponent;
+
+		before( 'Can log in', async function () {
```
I think it would be a good moment to decide on a test naming convention. Personally, I would drop the use of `Can` / `Should` / etc. and just go with the step-like description: 

`Log in` → `Start new post` → `Enter post title` → etc.

Also, I vaguely remember there were some issues with errors thrown from the before/after hooks, so it might be better to keep it as an `it` call. I might be mixing Jest with Mocha here though so calling @scinos for clarification! 🙏

worldomonation replied:
Agreed on the first point about naming convention for tests. Let's drop the `modal verbs`.

On the second point, in my experience so far there appears to be no issue with `mocha` when the before hook is placed in the test file. This placement ensures that `before` hook runs only once immediately preceding the inner `describe` block.

Whether to use the `before` hook was something I toyed with for some time before settling on the current approach, which is:
- if the login process is what is being tested (eg. wp-log-in-out spec) then use `it`.
- if the login process is not what is being tested (eg. likes) then use a hook to ensure it completes.

But as I typed out the previous bullets it occurred to me there is only a minor distinction between the two, really. So changing this to an `it` step could eliminate one consideration for test writers.

worldomonation replied:
Replaced `before` with `it` to treat the login step as just another test step, and removed modal verbs from step names.

WunderBart replied:
Thanks for elaborating! The tested vs not-tested distinction does make a lot of sense to me, haven't thought about it this way before. Sounds good to me whichever way it goes 👍

WunderBart commented on test/e2e/specs/specs-playwright/wp-likes-spec.js:
```diff
@@ -0,0 +1,56 @@
+
+		it( 'Like post', async function () {
+			likesComponent = await LikesComponent.Expect( this.page );
+			await likesComponent.clickLike();
```
How will we know the post was actually liked?

WunderBart replied:
Similar to [this](https://github.com/Automattic/wp-calypso/pull/53169/files#r639619316), would be nice here to have i.e. a `likesComponent.isLiked()` method to verify the state and add a necessary assertion.

worldomonation replied:
In the current implementation, I suppose there isn't an explicit check for if the request came back and acknowledged the click on the like button.

I assume the `Promise.race` in `clickLike` function serves as an implicit check on if the like button was indeed clicked, but it may not be enough. I will see if there's a way to make the check more explicit.

worldomonation replied:
So, this was bit of a struggle as I had the right idea, but each time had just _something_ wrong with my approach.

I was able to come up with a solution based on if the Like button has already been clicked or not, then checking for an event to be fired to the appropriate endpoint (which depend on if we're Liking or Unliking the post).

The Playwright logs confirm that my solution appears to work properly by detecting the request event.

Correct run
```
  pw:api   checking visibility of "text="You like this."" +2ms
  pw:api <= frame.isVisible succeeded +21ms
  pw:api => frame.click started +0ms
  pw:api => page.waitForEvent started +3ms   <--- wait for the request to a new like endpoint
  ...
  pw:api attempting click action +3ms
  pw:api   waiting for element to be visible, enabled and stable +0ms
  pw:api   element is visible, enabled and stable +16ms
  pw:api   scrolling into view if needed +0ms
  pw:api   done scrolling +1ms
  pw:api   checking that element receives pointer events at (316.22,748) +3ms
  pw:api   element does receive pointer events +3ms
  pw:api   performing click action +0ms
  pw:api   click action done +8ms
  pw:api   waiting for scheduled navigations to finish +0ms
  pw:api <= page.waitForEvent succeeded +1ms   <--- request resolved
  pw:api   navigations have finished +1ms
  pw:api <= frame.click succeeded +0ms
```

Intentionally incorrect run
```
  pw:api waiting for event "request" +1ms   <---  wait for request to endpoint that I intentionally mangled
  pw:api attempting click action +2ms
  pw:api   waiting for element to be visible, enabled and stable +0ms
  pw:api   element is visible, enabled and stable +6ms
  pw:api   scrolling into view if needed +1ms
  pw:api   done scrolling +0ms
  pw:api   checking that element receives pointer events at (316.22,748) +1ms
  pw:api   element does receive pointer events +2ms
  pw:api   performing click action +0ms
  pw:api   click action done +8ms
  pw:api   waiting for scheduled navigations to finish +1ms
  pw:api   navigations have finished +1ms
  pw:api <= frame.click succeeded +1ms
  pw:api   "load" event fired +2s
  pw:api   "domcontentloaded" event fired +1ms
  pw:api   "load" event fired +10ms
  pw:api   "networkidle" event fired +1s
  pw:api <= page.waitForEvent failed +11s  <--- event never resolved (as I would expect) leading to test fail.
```

WunderBart replied:
TIL that the request events are supported in Playwright, nice! 👍 It looks good, though we're relying on the request URLs now. I was thinking that we could maybe simplify it and just check if the button's state changed, which is how a user would confirm a successful action? Something like:

```js
async clickLike() {
  await this.frame.click( selectors.likeButton );
  await this.frame.waitForSelectort( selectors.likedButton );
}
```
What do you think?

worldomonation replied:
I simplified the logic and verified it still works 👍

WunderBart commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -0,0 +1,106 @@
+		await Promise.all( [
+			this.frame.click( this.snackBarNoticeLinkSelector ),
+			this.page.waitForNavigation(),
+			this.page.waitForLoadState( 'networkidle' ),
```
Should we make sure here that the post is actually viewed? I.e. by waiting until the "#page" element is visible? If yes, it's probably a good moment to init the `PostViewPage` component as well.

worldomonation replied:
The `waitForNavigation` only resolves once a navigation event has been detected. While that doesn't _guarantee_ that the navigation was for the page (as opposed to something else inadvertently causing it to navigate) I felt this was an acceptable assumption to make.

worldomonation replied:
I propose the implementation of PostViewPage or something similar in a follow-up so this PR does not grow any larger, if that is okay with you!

WunderBart commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -0,0 +1,106 @@
+
+	// Publish flow selectors.
+	firstPublishButtonSelector = '.editor-post-publish-panel__toggle';
+	secondPublishButtonSelector =
```
I'd keep those two above as `prePublishButtonSelector` and `publishButtonSelector`. Not a strong opinion here, just think that they describe the flow a bit better.

worldomonation replied:
I struggled with the naming of this a bit, and do not like neither the existing names nor my proposed names. There should be a better way to call them, perhaps by the location of where they are?

Perhaps `headerPublishButtonSelector` and `panelPublishButtonSelector`?

worldomonation replied:
I've renamed the selectors. Please let me know if you agree with the naming scheme - if not, I am happy to adopt your scheme.

WunderBart replied:
Naming is always the hardest part, I also struggle with it every time! 😄 
Maybe, if possible we should derive the name from the actual CSS selector, and if not then focus on the element function / type? Here's my take on these (notice the drop of `Selector` suffix as they're already under the `selectors` object):

```js
const selectors = {
	editorFrame: 'div.main.main-column.calypsoify.is-iframe > iframe',
	blockAppender: '.block-editor-default-block-appender',
	titleInput: '.editor-post-title__input',
	paragraphInput: 'p.block-editor-rich-text__editable:first-of-type',
	publishPanel: '.editor-post-publish-panel',
	publishPanelToggle: '.editor-post-publish-panel__toggle',
	publishButton: '.editor-post-publish-panel__header-publish-button button.editor-post-publish-button',
	viewPostButton: 'text=View Post',
};
```

Anyway, I'm not really strong on these but I would at least drop the `Selector` suffix 🗑️

worldomonation replied:
> I would at least drop the Selector suffix 🗑️

Good catch, it made sense when the selectors were in the class but now it's just repeating itself for no reason. Will push a revised commit today PDT.

With regard to names, let's codify that as the style guide going forward. I have plans to update the documentation with new Playwright style guide in #52464. I will push up a change to include name changes today.

worldomonation replied:
Implemented changes to selector names for the components in this PR. 

`LoginPage` still needs to migrate to the new paradigm.

WunderBart commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -0,0 +1,106 @@
+		await this.frame.click( this.firstPublishButtonSelector );
+		await this.frame.click( this.secondPublishButtonSelector );
+
+		await this.frame.waitForSelector( this.snackBarNoticeSelector );
```
I think it would be better to wait for the post-publish panel to appear here (`.editor-post-publish-panel`). [As mentioned earlier,](https://github.com/Automattic/wp-calypso/pull/53169#discussion_r639594748) the snackbar might not be a reliable element of validation.

WunderBart commented on test/e2e/specs/specs-playwright/wp-likes-spec.js:
```diff
@@ -0,0 +1,56 @@
+			gEditor = await GutenbergEditorPage.Expect( this.page );
+			const title = DataHelper.randomPhrase();
+			await gEditor.enterTitle( title );
+			await gEditor.enterText( quote );
```
Should we be checking if the title and quote are actually rendered in the editor canvas?

WunderBart replied:
I think it would be good to have two additional `GutenbergEditorPage` methods for verification: `getTitle` and `getText`. Then we could verify their presence and add proper assertions here.

worldomonation replied:
~~Sure, should not be difficult to do. Though Playwright would [throw an error if any of these fail](https://playwright.dev/docs/api/class-page#pagefillselector-value-options):~~

~~Only then is the element focused and filled.~~

~~I think adding a `frame.isVisible(selector)` should take care of checking on our end if text fill was successful.~~

I completely misunderstood what you were saying here. I will add the getter functions that can be used elsewhere in the tests or pages.

worldomonation replied:
Implemented `getText` and `getTitle` functions to permit verification of if the title and/or text was entered correctly.

This involved some investigation and frustration, surprisingly. It is noted in the comments in code, but the gist is:
- Playwright's `fill` function ignores newline characters, while `type` will enter as a human would, it is slow. I had to split the text up by newline, generate a list of promises to fill them one line at a time.
- in the process of getting the list of paragraph blocks to piece back together the text that is on page, I encountered the U+FEFF character which is present on empty paragraph blocks as a placeholder. This muddled up the string I was piecing back together.

I was able to overcome the issues for this test case, but we will see if my solution can stand up to other test cases.

WunderBart commented on test/e2e/specs/specs-playwright/wp-likes-spec.js:
```diff
@@ -0,0 +1,56 @@
+		} );
+
+		it( 'Enter post title', async function () {
+			gEditor = await GutenbergEditorPage.Expect( this.page );
```
Just a nitpick, but a strong one! 😄 I'd suggest a consistent naming for the instance objects. It's convenient to use shorter names but might bite us later on, when we have a gazillion different versions of it: `gEditor` / `gEditorPage` / `gutenbergEditor` / `gutenbergEditorPage` / etc. IMO it's just best to `snakeCase` the constructor and not even think about it: 
```js
const gutenbergEditorPage = await GutenbergEditorPage.Expect( this.page );
```

Maybe we can rename the `GutenbergEditorPage` to simply `GutenbergPage`? Are there any other pages we will be adding for Gutenberg?

worldomonation replied:
> Maybe we can rename the GutenbergEditorPage to simply GutenbergPage? Are there any other pages we will be adding for Gutenberg?

I would argue that we should maintain the naming of `GutenbergEditorPage` because it explicitly conveys that this page is an _Editor_ and it uses _Gutenberg_ for that purpose. It allows us to have additional editor types when the need arises.

worldomonation replied:
Renamed `gEditor` to the full `gutenbergEditorPage`.

WunderBart commented on test/e2e/specs/specs-playwright/wp-likes-spec.js:
```diff
@@ -0,0 +1,56 @@
+ */
+const mochaTimeOut = config.get( 'mochaTimeoutMS' );
+const host = DataHelper.getJetpackHost();
+const viewportName = BrowserHelper.getViewportName();
```
Looks like we'll be needing those for each suite (`mochaTimeOut`, `host`, `viewportName`). To limit the boilerplate, could we store those values in the suite instance as we currently do with the driver in Selenium, @scinos?

WunderBart replied:
See https://github.com/Automattic/wp-calypso/pull/53169#discussion_r639636628 which might reduce the boilerplate here.

worldomonation replied:
Selenium suites currently set this boilerplate at the beginning of each file. [Example](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/specs/specs-calypso/wp-invite-users__editor-spec.js#L29-L34).

Note that I agree it would be nice to remove the boilerplate. I suspect, but am not sure, that at least the `this.timeout` call needs to remain though, because otherwise `mocha` defaults to the 2000ms timeout, for some reason.

WunderBart commented on test/e2e/specs/specs-playwright/wp-likes-spec.js:
```diff
@@ -0,0 +1,56 @@
+	'The foolish man seeks happiness in the distance. The wise grows it under his feet.\n— James Oppenheim';
+
+describe( `[${ host }] Likes: (${ viewportName }) @parallel`, function () {
+	this.timeout( mochaTimeOut );
```
Are we still using mocha here?

worldomonation replied:
Yes, the Jest PR is still in draft state at this time.

With that said, I think it may be expedient to convert to Jest in a follow-up PR. This one is getting too large for my liking (typically I like a barrage of smaller PRs) if that's okay with you :)

worldomonation replied:
I've filed follow-up issue #53200 for the Jest migration.

WunderBart commented on test/e2e/specs/specs-playwright/wp-likes-spec.js:
```diff
@@ -0,0 +1,56 @@
+const quote =
+	'The foolish man seeks happiness in the distance. The wise grows it under his feet.\n— James Oppenheim';
+
+describe( `[${ host }] Likes: (${ viewportName }) @parallel`, function () {
```
Just an idea here - if we want to have all those top-level titles consistent, maybe we should have some kind of `createSuiteTitle` helper for that:

```js
// test-helper.js
import { browserHelper, dataHelper } from "@automattic/calypso-e2e";

const host = dataHelper.getJetpackHost();
const viewportName = browserHelper.getViewportName();

function createSuiteTitle( title ) {
  return `[${ host }] ${ title }: (${ viewportName }) @parallel`
}
```

```js
// wp-likes-spec.js
import { createSuiteTitle } from 'test-helper';

describe( createSuiteTitle( 'Likes' ), function () { 
  //
} );
```

worldomonation replied:
I'd like to hear what @scinos and @bsessions85 think about this.

On one hand, I like reducing boilerplate. It reduces inconsistencies and issues when copy-pasting into new files.
But on the other hand, having a function to generate test name would hide some meaningful information elsewhere in the codebase.

What do others think? Implementation wise, I am thinking it should be like `DataHelper.createSuiteTitle`.

WunderBart replied:
`DataHelper.createSuiteTitle` sounds good to me 👍

worldomonation replied:
I will file a follow-up task to address this.

bsessions85 replied:
This is an interesting approach. It is certainly something we could try and see how it looks. It would definitely create consistency.

worldomonation replied:
I've implemented the follow-up PR for the suite title: #53244 .

WunderBart commented on packages/calypso-e2e/src/lib/components/likes-component.ts:
```diff
@@ -0,0 +1,55 @@
+	 */
+	async clickLike(): Promise< void > {
+		await Promise.race(
+			[ 'text=Like', 'text=Liked' ].map( ( selector ) => this.frame.click( selector ) )
```
```suggestion
			[ this.likeButtonSelector, this.likedButtonSelector ].map( ( selector ) => this.frame.click( selector ) )
```

worldomonation replied:
Yes, my bad. Will be fixed in the next push.

WunderBart commented on packages/calypso-e2e/src/lib/components/likes-component.ts:
```diff
@@ -0,0 +1,55 @@
+	// Note the variation of 'Like'.
+	likeWidgetSelector = 'iframe.post-likes-widget';
+	likeButtonSelector = 'text=Like';
+	likedButtonSelector = 'text=Liked';
```
I think it would be better to define a separate `selectors` object. It would be then easier to reuse them as the IDE auto-complete would likely expand the list after typing `this.selectors.` (or just `selectors.`). Maybe something like this:

```js
const selectors = {
  likeWidget: 'iframe.post-likes-widget',
  likeButton: 'text=Like',
  likedButton: 'text=Liked',
};

export class LikesComponent extends BaseContainer {
  constructor( page: Page ) {
    super( page, selectors.likeWidget );
    this.selectors = selectors;
  }
  // 
}
```

What do you think?

worldomonation replied:
That is definitely an interesting idea. 

If the selector object is exported, then it also eliminates requirement for other pieces of code (that simply want to access a selector in that page/component) to instantiate an object which may be expensive time wise.

I will try this idea out. Thanks!

worldomonation replied:
Moved all selectors outside the class to a new object `selectors`.

worldomonation committed:
Address code review comments.
- split selectors off into its own object in the same file and replace existing references in all relevant files.
- discontinue use of snackbar for visiting published post in GutenbergEditorPage.
- correct duplicate selectors in Navbar.
- rename selectors in GutenbergEditorPage and standardize on terminology for sidebar vs pane.
- implement checks in the LikesComponent to verify the click had the intended effect.
- convert the in-suite `before` call to an `it`.

worldomonation committed:
Revise selector names to remove the '..selector' postfix.
Remove request monitoring logic in clickLike.

worldomonation committed:
Address PR comments.
- implement getText and getTitle functions in GutenbergEditorPage. This also required reworking the enterTitle and enterText functions.
- changed order and awaits for the NewPostFlow.

worldomonation committed:
Add debug env var and try out yet another way to click on the write button.

worldomonation committed:
Revert back to when the new post flow worked

worldomonation committed:
Update timeout to ms

worldomonation committed:
Shuffle around waits

worldomonation committed:
Fix selector in GutenbergEditorPage.

worldomonation committed:
Further shuffle around waits as the new post button isn't being clicked on TeamCity.

worldomonation commented:
There seems to be an issue with reliablly clicking the New Post button on the navbar/masterbar in the home screen that only occurs on TeamCity. Investigating why.

worldomonation commented:
Looks like the current approach of clicking on the New Post from the home screen works reliably now.

worldomonation committed:
Remove commented out const.

worldomonation committed:
Late night means unintended changes slipped in...

bsessions85 commented:
This is looking really good @worldomonation! I am realizing I am going to have to get used to using TS going forward!

worldomonation committed:
Implement a skeletal SidebarComponent and MyHomePage component, so the NewPostFlow can wait for presence of those elements before proceeding.

worldomonation commented:
> This is looking really good @worldomonation! I am realizing I am going to have to get used to using TS going forward!

Thanks! TypeScript isn't too difficult to pick up at all, though the transpiler and type checks are quite naggy.

worldomonation commented:
Turning back to draft yet again because of inconsistencies in, you guessed it, clicking the new post button in the navbar.

I suspect that Playwright is executing too fast for the test environment as I do not encounter the issue I am trying to address when running the test locally. 
On top of that, there may be something else eg. animation, element shift, invisible overlay that is interfering with Playwright's ability to click on the New Post button in a reliable manner.

With all this said, it is still quite surprising that it isn't able to reliably wait before clicking the New Post button, since that's one of the selling points of Playwright.

bsessions85 commented:
@worldomonation are you by chance hitting an a/b test? We need to make sure we're using the right user agent for the browser so that it gives us the default version in tests

worldomonation commented:
> @worldomonation are you by chance hitting an a/b test? We need to make sure we're using the right user agent for the browser so that it gives us the default version in tests

I don't think so - I do set the user agent [here](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/browser-manager.ts#L82) and IIRC the crucial part of the string (wp-e2e-tests) is there.

I made some more changes and have run the test a few times on TeamCity, and so far it seems to work. I believe I just needed to make it wait for both the sidebar and rest of the dashboard to populate, otherwise it executed way too fast.

kwight commented:
How can we test this locally? I'm assuming I have to make sure `mocha` is installed globally, but I'm also getting NPM errors about a missing package `Error: Cannot find module '@automattic/calypso-e2e'` (both on `npm i` and running the given `mocha` command).

worldomonation commented:
> How can we test this locally? I'm assuming I have to make sure `mocha` is installed globally, but I'm also getting NPM errors about a missing package `Error: Cannot find module '@automattic/calypso-e2e'` (both on `npm i` and running the given `mocha` command).

I have written a setup guide for running e2e tests are [here](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/docs/setup.md#software-environment) but I am guessing you already have everything checked out and dependencies installed locally.

As for the command, you should be able to run the test I added in this PR with:
`mocha --config .mocharc_playwright.yml specs/specs-playwright/wp-like-spec.js`

The `--config` parameter is important as otherwise mocha would load the default config file (`.mocharc.yml`) which is for the legacy Selenium based suite.

Alternatively, it is also possible to run my tests with `magellan`:
`yarn magellan --config=magellan-playwright.json --local_browser=chrome --test=specs/specs-playwright/wp-likes-spec.js`

Edit: I realized why you're running into that issue, @kwight .

You will need to first transpile TS to JS:

`yarn workspace @automattic/calypso-e2e build --watch`

worldomonation commented:
This PR is (again) exiting the draft stage and ready for review. I worked out the issue with `NewPostFlow` not reliably clicking the navbar button and I have not had issues on TeamCity with this PR (or any follow-up PRs based on this branch) for some time.

If this PR is too large, I am happy to break this up into two or more smaller PRs.
I say this because I prefer smaller, more numerous reviews and this specific PR is much larger than what I normally submit.

WunderBart commented:
> I am guessing that as per p1621762347012600/1621505612.010500-slack-C01QG4Y91RR I will need to ensure each of the last `await` calls do a `return await` instead?
> 
> To be honest, I am still a bit confused at the rule/expectation/whatever surrounding the `return await`. Whenever I run into issues with a failing test in Playwright, I do not have issues with the calls not being in the stack.

In the case of the `it()` callback there's no need to return anything. The rule regarding any helper though should be that **IF** the helper should return a certain value, and that value is being delivered by resolving a promise, the `return await` should be used [to keep the current context function in the error stack](https://github.com/Automattic/wp-calypso/pull/53145#discussion_r640626652). The additional benefit of that is the indication of an `async` function. Here's a bad & good example of such a helper:

**BAD**
```js
function getBodyText() {
  return this.driver.findElement( 'body' ).getText(); // returns a promise
}
```

**GOOD**
```js
async function getBodyText() {
  return await this.driver.findElement( 'body' ).getText(); // awaits the value and returns it
}
```

If the helper isn't expected to return any value, no return at the end is necessary as it won't have any effect on the error stack content.

We should be able to enforce this thanks to TS: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/return-await.md

cc @scinos

WunderBart commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -1,8 +1,24 @@
+	return this.replace( /\w\S*/g, function ( txt ) {
+		return txt.charAt( 0 ).toUpperCase() + txt.substr( 1 ).toLowerCase();
+	} );
+};
```
What's the benefit of extending the prototype vs creating a `properCase` utility (lodash-style)?

worldomonation replied:
No particular reason - I took what was in the Seleneium-based `dataHelper`. [This StackOverflow](https://stackoverflow.com/questions/8392035/add-method-to-string-class) also was the first hit when I searched `add method to string class` so I assumed extending the prototype is an acceptable approach.

Would you like me to use the alternative approach you suggested (creating utility with lodash)?

scinos replied:
+1 to create a separate utility (and not using `lodash` for it, see p4TIVU-9Bf-p2).
Extending the prototype is not very common anymore in JS projects.

WunderBart replied:
Thanks for the context, @worldomonation. I think it would be good to refactor it to the functional `properCase` utility. It should be a quick one so I think we can still do it within this PR 👍

worldomonation replied:
Thanks both for the context. I'll push up a change in the next PR.

WunderBart commented on packages/calypso-e2e/src/lib/components/likes-component.ts:
```diff
@@ -0,0 +1,80 @@
+			promises.push( this.frame.click( selectors.likeButton ) );
+		}
+
+		await Promise.all( promises );
```
This approach feels unintuitive to me as we need to declare events in reverse order. What are the benefits of calling `Promise.all` vs just `await` for individual steps?

worldomonation replied:
Yeah, this is one of the grips I have with Playwright - sometimes there is a need/suggestion to use `Promise.all`, and it isn't absolutely clear to someone relatively new to JS/TS (like myself) when one should use the `Promise.all` construct and when to use sequential `await` statements.

From what I have gleaned from countless hours on the documentation:

[Navigations](https://playwright.dev/docs/navigations) that may occur asynchronously or multiple times often require the use of `Promise.all` to prevent race condition between waiting for navigation and clicking on an element.

Popups often require the use of `Promise.all` to ensure the said popup is loaded to an acceptable state, though in a follow-up PR (yet to be posted) I have implemented popup handling without using `Promise.all`.

Ultimately, I think the use of `Promise.all` helps ensure that all required actions have been processed (in _some_ order) before proceeding. I think of it as a way to safeguard uncertain parts of the code, like clicking on the `Like` button that requires some delay after clicking for the request/response to be processed then reflected on the DOM.

With the (long wall of text) said, I can try to remove the need to use `Promise.all` here. You will likely see a lot more of `Promise.all`s coming in subsequent patches though, as they are really useful in Playwright.

worldomonation replied:
Well well well, guess sequential `awaits` can work, and seems reliable too. So much for all that headache trying to make the Promise.all work reliably.

WunderBart replied:
Thanks for the context, it explains a lot! I think using `Promise.all` makes sense in the navigation/popup scenario because we need to be already in the state of waiting for the **change event** before clicking the triggering link. In case of i.e. waiting for some element to appear on the current page as a result of **state change** (i.e. 'Like' button click), it should not be necessary to use `Promise.all`. In such a case, we're only checking if an element is present as a way of self-testing the helper, which should be fine as a simple sequential `await` call. Actually, I have a feeling that using `Promise.all` might be risky in some situations as it runs all the promises in parallel. It does fire them in given sequence, but it doesn't wait for resolution before calling the next one. See this example bin: https://jsbin.com/ropuyoj/1/edit?js,console

WunderBart commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -0,0 +1,166 @@
+		await Promise.all(
+			lines.map( async ( line, index ) => {
+				await this.frame.fill( `${ selectors.paragraphBlocks }:nth-of-type(${ index + 1 })`, line );
+				await this.page.keyboard.press( 'Enter' );
```
Nice!

worldomonation committed:
Try attempt to simplify the clickLike logic.

WunderBart commented on test/e2e/specs/specs-playwright/wp-likes-spec.js:
```diff
@@ -0,0 +1,64 @@
+			const title = DataHelper.randomPhrase();
+			await gutenbergEditorPage.enterTitle( title );
+			const enteredTitle = await gutenbergEditorPage.getTitle();
+			assert( title === enteredTitle );
```
I think it would be good to stay consistent in the aspect of self-validating our helpers. In the case of i.e. `clickLike` helper we are making sure that the action brings the expected result (the "Like" button turns into the "Liked" button). Here, we shift that responsibility to the test itself. This could lead to a situation where this helper needs to be validated every time it's used in a test. What do you think? Would this be something we could consider adding to the documentation?

worldomonation replied:
My answer here to the question of `self-validating our helpers` is going to be a 'sometimes, maybe'. 

[This SQA post](https://sqa.stackexchange.com/questions/11624/include-assertions-expectations-within-the-page-objects-yes-or-no) is a bit old, but it outlines that some assertions are better served in the page object while others are better in the tests.

[Another useful SQA post](https://sqa.stackexchange.com/questions/27751/my-understanding-are-correct-or-not-in-page-object-pattern/27765) references the `Arrange, Act, Assert` principle which advocates for assertions to be primarily in the tests.

So, in the end it seems to be a mixed bag. I felt there was a distinction between the title assertion and clickLike assertion, but maybe they are two sides of the same coin after all.

----
What if I move the assertion inside `GutenbergEditorPage.getTitle`, would it feel more consistent?

worldomonation replied:
I've restructured the API a bit.

`GutenbergEditorComponent.enterTitle()` now calls two distinct methods that act as a setter/getter. The setter sets the title which is then retrieved by the getter, and assertion takes place in the `enterTitle` method call before control is passed back to the caller.

This way, the setter is only concerned with filling in the text.

WunderBart commented on packages/calypso-e2e/src/lib/pages/my-home-page.ts:
```diff
@@ -0,0 +1,24 @@
+	}
+
+	async _postInit(): Promise< void > {
+		await this.page.waitForSelector( selectors.statsCard );
```
Is it necessary to wait for the additional `.stats` element here? I can't see it in the [current `MyHomePage` component](https://github.com/Automattic/wp-calypso/blob/bd262f39a92e04d391bcfd0665e7dc8176f66516/test/e2e/lib/pages/my-home-page.js).

worldomonation replied:
In my experience so far, calypso home dashboard is a bit weird like that. 

Playwright offers three load states to look out for:
- `load` - fires the earliest
- `domcontentloaded` - fires once the HTML document is loaded
- `networkidle` - when no network activity occurs for 500ms

I had thought waiting for `networkidle` was enough to ensure the home dashboard loaded entirely, but this is not the case. 

Waiting for the stats graph card is my alternative solution to make Playwright wait a bit more before rushing off to click on the navbar write post button. I added it in [this commit](https://github.com/Automattic/wp-calypso/pull/53169/commits/865f56bebbe443df219530dd10d311db1f50b450); prior to this change, every commit would fail between the login step and waiting for the editor to appear.

Now, on the local machine - this wait isn't strictly necessary since local hardware typically runs much faster. This is only an issue on TeamCity where the machines don't run quite as fast.

worldomonation replied:
I've removed the wait for `statsCard` to load since merely waiting for the Dashboard appears to take care of the loading issue.

WunderBart commented on test/e2e/specs/specs-playwright/wp-likes-spec.js:
```diff
@@ -0,0 +1,64 @@
+		it( 'Enter post text', async function () {
+			await gutenbergEditorPage.enterText( quote );
+			const enteredText = await gutenbergEditorPage.getText();
+			assert( quote === enteredText );
```
Same as https://github.com/Automattic/wp-calypso/pull/53169/files#r643910584

worldomonation committed:
Move the assertion of title and text to the library methods.

worldomonation committed:
Try removing the wait for stats card in MyHomePage.

worldomonation committed:
Discontinue extending prototype and instead create a utility function within DataHelper.

worldomonation committed:
Remove wait for chart to load in MyHomePage.

scinos commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -60,12 +65,26 @@ export class GutenbergEditorPage extends BaseContainer {
+		const sanitizedTitle = title.trim();
+		await this.setTitle( sanitizedTitle );
+		const readBack = await this.getTitle();
+		assert( readBack === sanitizedTitle );
```
`assert.strictEqual(readBack, sanitizedTitle)` will give you a better error message in case it fails

scinos reviewed (approved):
Approved from the platform point of view. I don't have enough knowledge to evaluate the usage of Playwright APIs, the components or the spec, sorry :(

WunderBart reviewed (approved):
Looks good, thanks! 🚀

worldomonation commented:
> Approved from the platform point of view. 

Thanks for the review!

> I don't have enough knowledge to evaluate the usage of Playwright APIs, the components or spec

That's perfectly okay as I am primarily looking for feedback and critique on code style, how I have implemented something and suggestions on better approaches. I've found a lot of value in your critique of how I've implemented something as usually there's a better way than what I've come up with.
